### PR TITLE
feat: adding artifact version deletion api

### LINKED ISSUE #3074: Support for artifact version deletion
Consider adding a functionality that allows deleting an artifact version. This is related to this [conversation](https://github.com/Apicurio/apicurio-registry/issues/363#issuecomment-1346807360).

### PR DESCRIPTION
**Pull request by derickmr**
Resolves #3074

### CODE REVIEW

File: app/src/main/java/io/apicurio/registry/rest/RestConfig.java [hunk 1]
```diff
@@ -17,8 +17,16 @@ public class RestConfig {
     @Info(category = "rest", description = "Skip SSL validation when downloading artifacts from URL", availableSince = "2.2.6-SNAPSHOT")
     boolean downloadSkipSSLValidation;
 
+    @ConfigProperty(name = "registry.rest.artifact.deletion.enabled", defaultValue = "true")
+    @Info(category = "rest", description = "Enables artifact version deletion", availableSince = "2.4.2-SNAPSHOT")
+    boolean artifactVersionDeletionEnabled;
```
Comment (jsenko) [inline app/src/main/java/io/apicurio/registry/rest/RestConfig.java:22]: We should use dynamic configuration property here, with a disclaimer in the description, so it works in a multi-tenant environment. Similarly to enabling basic auth.

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java [hunk 2]
```diff
@@ -564,6 +565,22 @@ public Response getArtifactVersion(String groupId, String artifactId, String ver
...
+    public void deleteArtifactVersion(String groupId, String artifactId, String version) {
+        if (!restConfig.isArtifactVersionDeletionEnabled()) {
+            throw new NotFoundException("Artifact version deletion is not enabled.");
+        }
+
+        requireParameter("groupId", groupId);
+        requireParameter("artifactId", artifactId);
+        requireParameter("version", version);
+
+        storage.deleteArtifactVersion(gidOrNull(groupId), artifactId, version);
```
Comment (jsenko) [inline app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:581]: We should think about and handle the following edge cases:
- What happens if the version is the latest version
- What happens if the version (or other versions) is in a `DISABLED` state
- What happens if the version is he only one in the artifact
  - We should require deleting the entire artifact instead

File: app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json [hunk 3]
```diff
@@ -992,6 +992,25 @@
                 "summary": "Get artifact version",
                 "description": "Retrieves a single version of the artifact content.  Both the `artifactId` and the\nunique `version` number must be provided.  The `Content-Type` of the response depends \non the artifact type.  In most cases, this is `application/json`, but for some types \nit may be different (for example, `PROTOBUF`).\n\nThis operation can fail for the following reasons:\n\n* No artifact with this `artifactId` exists (HTTP error `404`)\n* No version with this `version` exists (HTTP error `404`)\n* A server error occurred (HTTP error `500`)\n"
             },
+            "delete": {
+                "tags": [
+                    "Versions"
+                ],
+                "responses": {
```
Comment (jsenko) [inline app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json:999]: What would the response look like if the feature is disabled?

File: app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json [hunk 4]
```diff
@@ -992,6 +992,25 @@
...
+                    "404": {
+                        "$ref": "#/components/responses/NotFound"
+                    },
+                    "500": {
+                        "$ref": "#/components/responses/ServerError"
+                    }
+                },
+                "operationId": "deleteArtifactVersion",
+                "summary": "Delete artifact version",
+                "description": "Deletes a single version of the artifact.  Both the `artifactId` and the unique `version`\nare needed.  If this is the only version of the artifact, this operation is the same as \ndeleting the entire artifact.\n\nThis operation can fail for the following reasons:\n\n* No artifact with this `artifactId` exists (HTTP error `404`)\n* No version with this `version` exists (HTTP error `404`)\n* A server error occurred (HTTP error `500`)\n"
```
Comment (jsenko) [inline app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json:1012]: We need to add information about how to enable this feature, and that it's discouraged for normal use.

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java [hunk 5]
```diff
@@ -564,6 +565,22 @@ public Response getArtifactVersion(String groupId, String artifactId, String ver
         return builder.build();
     }
 
+    /**
+     * @see io.apicurio.registry.rest.v2.GroupsResource#deleteArtifactVersion(java.lang.String, java.lang.String, java.lang.String)
+     */
+    @Override
```
Comment (jsenko) [inline app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:571]: Authorization annotation is missing. We thing it should be `AuthorizedLevel.Write`.

File: app/src/main/resources/application.properties [hunk 6]
```diff
@@ -180,3 +180,6 @@ quarkus.native.additional-build-args=--initialize-at-run-time=org.apache.kafka.c
 # Artifact download
 registry.rest.artifact.download.maxSize=1000000
 registry.rest.artifact.download.skipSSLValidation=false
+
+# Artifact version deletion
+registry.rest.artifact.deletion.enabled=true
```
Comment (jsenko) [inline app/src/main/resources/application.properties:185]: Should be disabled by default.

File: app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java [hunk 7]
```diff
@@ -636,6 +636,87 @@ public void testGetArtifactMetaDataByContent() throws Exception {
...
+            if (idx % 2 == 0) {
+                continue;
+            }
+
+            Integer version = versions.get(idx);
+            given()
+                .when()
+                    .pathParam("artifactId", "testDeleteArtifactVersion/EmptyAPI")
+                    .pathParam("version", version)
+                    .delete("/registry/v1/artifacts/{artifactId}/versions/{version}")
```
Comment (jsenko) [inline app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java:677]: This doesn't seem right, does the v1 API support version deletion?

Comment (jsenko) [inline app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java:677]: This doesn't seem right, does the v1 API support version deletion?

**Comment by jsenko** [comment]
Hi @derickmr , thank you for the PR. We had a discussion and we agreed that we want to include this feature. I've added a few comments that came up. We'd be happy if you help with some of the changes in the comments, but we can also merge the PR and make additional changes afterwards. Let us what you think!

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java (hunk 2 above)
Comment (jsenko) [reply app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:581]: + tests for the edge cases

File: app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java (hunk 7 above)
Comment (EricWittmann) [reply app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java:677]: We definitely shouldn't be adding Delete Version to the v1 API.  Adding it to v2 makes sense, but v1 is frozen.  And I don't think v1 ever supported version delete, unless I'm mis-remembering.

File: app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json (hunk 3 above)
Comment (EricWittmann) [reply app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json:999]: Good question.  What are the choices?  

* `401` - Authorization Failed
* `403` - Forbidden
* `405` - Method Not Allowed
* `406` - Not Acceptable
* `409` - Conflict

None of those map all that well onto "Sorry, that feature is currently disabled".

I think `403` and `405` are probably the best candidates, with `405` probably getting my vote.

File: app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java (hunk 7 above)
Comment (derickmr) [reply app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java:677]: Oops, sorry, my mistake. This code was supposed to be removed

File: app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json (hunk 3 above)
Comment (derickmr) [reply app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json:999]: I think `405` sounds like a reasonable choice for me, but let me know your thoughts

File: app/src/main/java/io/apicurio/registry/rest/RestConfig.java (hunk 1 above)
Comment (derickmr) [reply app/src/main/java/io/apicurio/registry/rest/RestConfig.java]: Fixed

File: app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json (hunk 3 above)
Comment (derickmr) [reply app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json:999]: Implemented 405 response for the case

File: app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json (hunk 4 above)
Comment (derickmr) [reply app/src/main/resources-unfiltered/META-INF/resources/api-specifications/registry/v2/openapi.json]: Fixed

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java (hunk 5 above)
Comment (derickmr) [reply app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:572]: Fixed

File: app/src/main/resources/application.properties (hunk 6 above)
Comment (derickmr) [reply app/src/main/resources/application.properties]: Fixed

File: app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java (hunk 7 above)
Comment (derickmr) [reply app/src/test/java/io/apicurio/registry/noprofile/rest/v1/ArtifactsResourceTest.java]: Fixed

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java (hunk 2 above)
Comment (derickmr) [reply app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java]: Hi @jsenko, based on the code I see on AbstractSqlRegistryStorage#deleteArtifactVersion:
- If the latest version is deleted, then the latest version is re-assigned
image

- Looks like there's no state validation, meaning the latest can be re-assigned to a disabled version

- If artifact has only one version, then the whole artifact is deleted
image

This method has tests covered under AbstractRegistryStorageTest#testDeleteArtifactVersion. I see there is test that cover deleting the latest version:
image

I can help increasing test coverage for these edge cases and perhaps add some kind of validation at api level, but I am reluctant on changing the internal functionality itself as I am not fully aware of how this could impact other areas of the code - would be more efficient if these eventual internal functionality changes is handled by Apicurio team I believe. 

Please let me know your thoughts

**Comment by derickmr** [comment]
> Hi @derickmr , thank you for the PR. We had a discussion and we agreed that we want to include this feature. I've added a few comments that came up. We'd be happy if you help with some of the changes in the comments, but we can also merge the PR and make additional changes afterwards. Let us know what you think!

Hi @jsenko, I covered all comments. I left a reply on your comment related to edge cases. Let me know your thoughts. Thanks for giving attention to this one!

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java (hunk 2 above)
Comment (derickmr) [reply app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java]: Btw, added some more coverage to AbstractSqlRegistryStorage#deleteArtifactVersion to help covering those edge-case scenarios

**Comment by derickmr** [comment]
Hi @jsenko, please let me know your thoughts on latest changes.

**Comment by derickmr** [comment]
Hi @EricWittmann, @jsenko, please let me know if need anything else from my side to get another review on this PR. Comments from last review were addressed already. Thank you

**Comment by EricWittmann** [comment]
Yeah sorry about this.  Putting it on the review schedule for Monday.

**Review by jsenko — APPROVED** [review]
